### ISSUE TITLE
Clarification on outcomes for tests for rule `921140` 

### ISSUE THREAD

**Issue by TimDiam0nd** [coreruleset/coreruleset#3824]
Hi

I am struggling to understand the outcomes section for the two tests for rule id [921140](https://github.com/coreruleset/coreruleset/blob/1712513bb22de44fed74ee302849b2740395b704/rules/REQUEST-921-PROTOCOL-ATTACK.conf#L121). 

Essentially, the rule looks for either a `\n` or a `\r` (`rx [\n\r]`) in any of the request headers or header names. 

The two tests contain:
https://github.com/coreruleset/coreruleset/blob/1712513bb22de44fed74ee302849b2740395b704/tests/regression/tests/REQUEST-921-PROTOCOL-ATTACK/921140.yaml#L16
which contains a `\r`, and is correctly quoted in yaml to be treated as a literal `\r` 
and
https://github.com/coreruleset/coreruleset/blob/1712513bb22de44fed74ee302849b2740395b704/tests/regression/tests/REQUEST-921-PROTOCOL-ATTACK/921140.yaml#L33
which contains `%0d`, which is decoded to a `\r` by the html entity decode transformation function.

Now, based on the above, both tests should pass. I cannot see a specific reason as to why these tests should be set to `no_expect_ids` as currently is:
https://github.com/coreruleset/coreruleset/blob/1712513bb22de44fed74ee302849b2740395b704/tests/regression/tests/REQUEST-921-PROTOCOL-ATTACK/921140.yaml#L23 and
https://github.com/coreruleset/coreruleset/blob/1712513bb22de44fed74ee302849b2740395b704/tests/regression/tests/REQUEST-921-PROTOCOL-ATTACK/921140.yaml#L40

Is this due to something specific in Apache? If so, is it possible if something can be added to the test description explaining this?

To add to this, the next rule, [921150](https://github.com/coreruleset/coreruleset/blob/1712513bb22de44fed74ee302849b2740395b704/rules/REQUEST-921-PROTOCOL-ATTACK.conf#L149) is very similar in that it also looks for a `\n` or `\r`, but in query args instead. The two tests for that rule both rightly contain `expect_ids` as expected.

**Comment by theseion** [comment]
Well spotted. The rule looks for characters in the _request_ headers that can cause request splitting in the _response_.  There are a couple of issues here:
1. `\r` ends up as the literal `\r` in the header name. I don't know whether it's ModSecurity of Apache that causes this. This is the issue that you're seeing, because the literal isn't detected by the rule.
2. Looking at the second test, the rule should also detect the literals `%0d` and `%0a`, as the expectation would be that an application would decode them and then cause splitting of the response
3. There should be tests for `\n` and `%0a` as well
4. `\n` will cause a "Bad Request" in Apache. This is fine but the test would need to be written accordingly
5. For all tests, nginx and Coraza need to be checked and potentially overriden in the "override-*" files

**Comment by theseion** [comment]
@airween can you tell why the `\r` ends up as the `\r` literal instead of the actual byte value in the header? `\n` is interpreted as expected and causes a `400` in Apache.

**Comment by airween** [comment]
> @airween can you tell why the `\r` ends up as the `\r` literal instead of the actual byte value in the header? `\n` is interpreted as expected and causes a `400` in Apache.

I'm not sure `\r` in yaml will be interpreted as `chr(13)` - few days/weeks ago I asked a similar question on Slack (not about `\r` but about `chr(14)`), and we agreed that in these case we use `base64` encoded request.

But first we should check how `\r` behaves in a `go-ftw` generated request. I try to look at that soon.

**Comment by airween** [comment]
Okay, I checked that and seems like `go-ftw` sends `\r` as literal, not a control character.

Run the test:
```
./ftw run --config .ftw.apache.yaml -d ../coreruleset/tests/regression/tests/ --include 921140-1
```
Capture the traffic:
```
tcpdump -i lo src port 80 or dst port 80 -XX
```
See the (relevant) content:
```
	0x00d0:  0d0a 486f 7374 3a20 6c6f 6361 6c68 6f73  ..Host:.localhos
	0x00e0:  740d 0a53 6f6d 6548 6561 6465 723a 2048  t..SomeHeader:.H
	0x00f0:  6561 6465 7264 6174 615c 7249 6e6a 6563  eaderdata\rInjec
	0x0100:  7465 6448 6561 6465 723a 2072 6573 706f  tedHeader:.respo
	0x0110:  6e73 655f 7370 6c69 7474 696e 675f 636f  nse_splitting_co
	0x0120:  6465 0d0a 5573 6572 2d41 6765 6e74 3a20  de..User-Agent:.
	0x0130:  4f57 4153 5020 4352 5320 7465 7374 2061  OWASP.CRS.test.a
	0x0140:  6765 6e74 0d0a 0d0a                      gent....
```
In 3rd line there is the `data\rInjec` sub-string, and the bytes:
```
...5c 72...
```
```python
>>> print("%c%c" % (chr(0x5c), chr(0x72)))
\r
```
So this is literally 2 bytes: `\` and `r`.

**Comment by Xhoenix** [comment]
The `no_expect_ids` are correct, as the rule doesn't gets triggered. But, as per the rule modsec should detect `%0d%0a`, but it doesn't, which is an issue, and leads to a false negative situation.

**Comment by theseion** [comment]
Thanks @airween. I'll have to check go-ftw then. My expectation for Apache is that sending `\r` in a header will also cause a status `400`. In any case, `no_expect_ids` isn't correct.

**Comment by airween** [comment]
> Thanks @airween. I'll have to check go-ftw then. My expectation for Apache is that sending `\r` in a header will also cause a status `400`. In any case, `no_expect_ids` isn't correct.

Yes, you're definitely right: if a header contains a `\r` (`chr(0x0d)`) character then Apache responses with `HTTP 400`.
[421140.txt](https://github.com/user-attachments/files/17088982/421140.txt)

See the test with the attached file:
```
cat 421140.txt | netcat 127.0.0.1 80
```

**Comment by theseion** [comment]
I don't think so. URL decoding doesn't make sense either but it's at least conceivable. If an application performs HTML entity decoding of an HTTP header, well, then nothing can save it :)

**Comment by theseion** [comment]
Unfortunately, processing of `\r` in double quoted YAML strings is a bug in the YAML library we are using in go-ftw. I've opened an issue in the [go-ftw repo](https://github.com/coreruleset/go-ftw/issues/359). However, we are currently blocked from upgrading that library due to another issue.

**Comment by theseion** [comment]
Let's put this issue on hold until we can fix go-ftw.

**Comment by TimDiam0nd** [comment]
> I don't think so. URL decoding doesn't make sense either but it's at least conceivable. If an application performs HTML entity decoding of an HTTP header, well, then nothing can save it :)

Yep apologies, i was referencing the transformation functions being applied, in that if we have a URL encoded `\r`, but only htmlEntityDecode trans func, the rule will never match the test. If url encoded values are to be matched, the urlDecode transformation func should be added to the rule right?

**Comment by theseion** [comment]
> Yep apologies, i was referencing the transformation functions being applied, in that if we have a URL encoded \r, but only htmlEntityDecode trans func, the rule will never match the test. If url encoded values are to be matched, the urlDecode transformation func should be added to the rule right?

You are right, logically. But URL decoding would be something that an application applies to a header, so CRS shouldn't care about it. I think the test just doesn't make sense in this instant.

**Comment by theseion** [comment]
Check again after [removing goccy/go-yaml](https://github.com/coreruleset/go-ftw/pull/375).

**Comment by theseion** [comment]
The current YAML library correctly unmarshals escape sequences in double quoted string scalars. Now we need to properly look at the rules again. To my surprise, they really appear to guard against application level decoding of header names and values.

**Comment by theseion** [comment]
@franbuehler @dune73 921140 also deals with double URL encoding. Would appreciate your input here.

**Comment by franbuehler** [comment]
If we change the transformation function from `t:htmlEntityDecode` to `t:urlDecodeUni` in rule 921140, then at least the [second test](https://github.com/coreruleset/coreruleset/blob/main/tests/regression/tests/REQUEST-921-PROTOCOL-ATTACK/921140.yaml#L33) triggers the rule 921140.

The rule [921140](https://github.com/coreruleset/coreruleset/blob/1712513bb22de44fed74ee302849b2740395b704/rules/REQUEST-921-PROTOCOL-ATTACK.conf#L121) is similar to [921151](https://github.com/coreruleset/coreruleset/blob/main/rules/REQUEST-921-PROTOCOL-ATTACK.conf#L321), [921150](https://github.com/coreruleset/coreruleset/blob/main/rules/REQUEST-921-PROTOCOL-ATTACK.conf#L149) and [921190](https://github.com/coreruleset/coreruleset/blob/main/rules/REQUEST-921-PROTOCOL-ATTACK.conf#L197). They all have similar tests with %0a, except [one test in 921151](https://github.com/coreruleset/coreruleset/blob/main/rules/REQUEST-921-PROTOCOL-ATTACK.conf#L321) that also has a `no_expect_ids`.

For me, it seems rule 921140 is working as intended, isn't it?

**Comment by theseion** [comment]
Thanks @franbuehler, I completely agree.

**Comment by vagrantsd** [comment]
Hi.
After updating the rules to version 4.12.0.
I got a lot of false positives. And selectively. from some browsers the same url worked without blocks. And other browsers had problems and triggered 949110 
And when opening the root page of my resource.
after I returned htmlEntityDecode and removed urlDecodeUni in the rules.
921140
921150
921160
921151
The problem disappeared.
Where could be the bug and how to find it. Is it in urlDecodeUni or something in my resource code?

**Comment by theseion** [comment]
@vagrantsd I've moved your question to a new issue here: https://github.com/coreruleset/coreruleset/issues/4059.
